File: poc/atproto/dds-reference/python/dds_client/dds_client/feed.py
"""Feed consumer. Mirrors `packages/client-ts/src/feed.ts`.

`subscribe()` yields DDS records via Jetstream, optionally preceded by a PDS catch-up
pass for the requested (DID, collection) pairs.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from datetime import datetime
from typing import AsyncIterator, Optional

import httpx
import websockets

from .version import COLLECTIONS


@dataclass
class DDSFeedEvent:
    source: str  # "jetstream" | "catchup"
    did: str
    collection: str
    rkey: str
    uri: str
    cid: Optional[str]
    time_ms: float
    record: dict


JETSTREAM_DEFAULT = "wss://jetstream1.us-east.bsky.network/subscribe"
APPVIEW_DEFAULT = "https://bsky.social"


async def _catchup(
    *,
    since: datetime,
    dids: list[str],
    collections: list[str],
    service: str,
    logger,
) -> AsyncIterator[DDSFeedEvent]:
    since_ms = since.timestamp() * 1000
    async with httpx.AsyncClient(base_url=service, timeout=30.0) as http:
        for did in dids:
            for collection in collections:
                cursor: Optional[str] = None
                while True:
                    params = {"repo": did, "collection": collection, "limit": 100}
                    if cursor:
                        params["cursor"] = cursor
                    try:
                        resp = await http.get("/xrpc/com.atproto.repo.listRecords", params=params)
                        resp.raise_for_status()
                        body = resp.json()
                    except Exception as exc:
                        logger(f"catchup error did={did} collection={collection}: {exc}")
                        break
                    for r in body.get("records", []):
                        val = r.get("value") or {}
                        created = val.get("createdAt")
                        try:
                            ts = datetime.fromisoformat(created.replace("Z", "+00:00")).timestamp() * 1000
                        except Exception:
                            continue
                        if ts < since_ms:
                            continue
                        uri = r["uri"]
                        parts = uri.replace("at://", "").split("/", 2)
                        if len(parts) != 3:
                            continue
                        yield DDSFeedEvent(
                            source="catchup",
                            did=parts[0],
                            collection=parts[1],
                            rkey=parts[2],
                            uri=uri,
                            cid=r.get("cid"),
                            time_ms=ts,
                            record=val,
                        )
                    cursor = body.get("cursor")
                    if not cursor:
                        break


async def subscribe(
    *,
    collections: Optional[list[str]] = None,
    dids: Optional[list[str]] = None,
    cursor_us: Optional[int] = None,
    catchup_since: Optional[datetime] = None,
    catchup_service: str = APPVIEW_DEFAULT,
    jetstream_endpoint: str = JETSTREAM_DEFAULT,
    logger=lambda msg: None,
    stop_event: Optional[asyncio.Event] = None,
) -> AsyncIterator[DDSFeedEvent]:
    """Yield DDS records, catch-up first then live via Jetstream.

    Set `stop_event.set()` to cleanly stop the iterator.
    """
    cols = collections or COLLECTIONS.all()

    if catchup_since and dids:
        async for ev in _catchup(
            since=catchup_since, dids=dids, collections=cols, service=catchup_service, logger=logger
        ):
            yield ev

    params = []
    for c in cols:
        params.append(("wantedCollections", c))
    for d in dids or []:
        params.append(("wantedDids", d))
    if cursor_us is not None:
        params.append(("cursor", str(cursor_us)))
    qs = "&".join(f"{k}={v}" for k, v in params)
    url = f"{jetstream_endpoint}?{qs}" if qs else jetstream_endpoint

    logger(f"DDS feed: connecting to {url}")
    async for ws in websockets.connect(url, max_size=None):
        try:
            while not (stop_event and stop_event.is_set()):
                msg = await ws.recv()
                event = json.loads(msg)
                if event.get("kind") != "commit":
                    continue
                commit = event.get("commit") or {}
                if commit.get("operation") not in ("create", "update"):
                    continue
                rec = commit.get("record")
                if not isinstance(rec, dict):
                    continue
                did = event["did"]
                collection = commit["collection"]
                rkey = commit["rkey"]
                yield DDSFeedEvent(
                    source="jetstream",
                    did=did,
                    collection=collection,
                    rkey=rkey,
                    uri=f"at://{did}/{collection}/{rkey}",
                    cid=commit.get("cid"),
                    time_ms=event.get("time_us", 0) / 1000,
                    record=rec,
                )
            if stop_event and stop_event.is_set():
                await ws.close()
                return
        except websockets.ConnectionClosed:
            if stop_event and stop_event.is_set():
                return
            logger("DDS feed: connection closed, reconnecting...")
            continue

File: poc/atproto/dds-reference/packages/client-ts/src/feed.ts
import { Jetstream } from "@skyware/jetstream";
import { Agent } from "@atproto/api";
import { COLLECTIONS, type DDSCollection } from "./version.js";

/** A normalized event surfaced by `subscribe`. Wraps both live jetstream events and PDS catch-up. */
export interface DDSFeedEvent {
  /** Source of this event: live jetstream stream, or PDS catch-up. */
  source: "jetstream" | "catchup";
  /** AT Protocol repo DID this record belongs to. */
  did: string;
  /** Collection NSID. */
  collection: DDSCollection | string;
  rkey: string;
  uri: string;
  cid?: string;
  /** Timestamp normalized to milliseconds since epoch. */
  timeMs: number;
  /** The record value as authored. */
  record: Record<string, unknown>;
}

export interface DDSSubscribeOptions {
  /** Which DDS collections to follow. Defaults to all v1 collections. */
  collections?: string[];
  /** Specific DIDs to filter by. */
  dids?: string[];
  /** Optional cursor (unix microseconds, jetstream-native) to resume from. */
  cursor?: number;
  /**
   * If set, the consumer will, on startup, call `com.atproto.repo.listRecords` against each
   * `dids[]` x `collections[]` pair, emitting any record newer than `catchupSince` as a
   * `source: "catchup"` event before live streaming starts. Use this to fill gaps after downtime.
   */
  catchupSince?: Date;
  /**
   * If provided, used to fetch PDS records during catch-up. If omitted, an unauthenticated
   * Agent against `bsky.social` is used. Catch-up requires `dids` set.
   */
  agent?: Agent;
  /** Optional logger; defaults to a no-op. */
  logger?: { info: (msg: string) => void; error: (err: unknown) => void };
  /** AbortSignal to cleanly stop the consumer. */
  signal?: AbortSignal;
}

const NOP_LOGGER = { info: (_: string) => {}, error: (_: unknown) => {} };

/**
 * Subscribe to DDS records as an async iterator. Yields catch-up records first (if requested),
 * then live jetstream events. Stops when the signal aborts.
 */
export async function* subscribe(
  opts: DDSSubscribeOptions = {},
): AsyncGenerator<DDSFeedEvent, void, undefined> {
  const collections = opts.collections ?? Object.values(COLLECTIONS);
  const logger = opts.logger ?? NOP_LOGGER;

  if (opts.catchupSince && opts.dids?.length) {
    yield* runCatchup({
      since: opts.catchupSince,
      dids: opts.dids,
      collections,
      agent: opts.agent ?? new Agent("https://bsky.social"),
      logger,
      signal: opts.signal,
    });
  }

  const jetstream = new Jetstream({
    wantedCollections: collections,
    wantedDids: opts.dids,
    cursor: opts.cursor,
  });

  const queue: DDSFeedEvent[] = [];
  let waiter: ((ev: DDSFeedEvent | null) => void) | null = null;

  const push = (ev: DDSFeedEvent) => {
    if (waiter) {
      const w = waiter;
      waiter = null;
      w(ev);
    } else {
      queue.push(ev);
    }
  };

  jetstream.on("commit", (event) => {
    const commit = event.commit;
    if (commit.operation !== "create" && commit.operation !== "update") return;
    const rec = (commit as { record?: unknown }).record;
    if (!rec || typeof rec !== "object") return;
    push({
      source: "jetstream",
      did: event.did,
      collection: commit.collection,
      rkey: commit.rkey,
      uri: `at://${event.did}/${commit.collection}/${commit.rkey}`,
      cid: (commit as { cid?: string }).cid,
      timeMs: event.time_us / 1000,
      record: rec as Record<string, unknown>,
    });
  });

  jetstream.on("error", (err) => logger.error(err));

  const onAbort = () => {
    jetstream.close();
    if (waiter) {
      const w = waiter;
      waiter = null;
      w(null);
    }
  };
  opts.signal?.addEventListener("abort", onAbort, { once: true });

  jetstream.start();
  logger.info("DDS feed: jetstream started");

  try {
    while (!opts.signal?.aborted) {
      if (queue.length > 0) {
        yield queue.shift()!;
        continue;
      }
      const ev = await new Promise<DDSFeedEvent | null>((resolve) => {
        waiter = resolve;
      });
      if (ev === null) break;
      yield ev;
    }
  } finally {
    opts.signal?.removeEventListener("abort", onAbort);
    jetstream.close();
    logger.info("DDS feed: jetstream stopped");
  }
}

async function* runCatchup(args: {
  since: Date;
  dids: string[];
  collections: string[];
  agent: Agent;
  logger: NonNullable<DDSSubscribeOptions["logger"]>;
  signal: AbortSignal | undefined;
}): AsyncGenerator<DDSFeedEvent, void, undefined> {
  const sinceMs = args.since.getTime();
  for (const did of args.dids) {
    for (const collection of args.collections) {
      if (args.signal?.aborted) return;
      let cursor: string | undefined;
      do {
        const res = await args.agent.com.atproto.repo
          .listRecords({ repo: did, collection, limit: 100, cursor })
          .catch((err) => {
            args.logger.error(err);
            return undefined;
          });
        if (!res) break;
        for (const r of res.data.records) {
          const value = r.value as Record<string, unknown> | null;
          if (!value || typeof value !== "object") continue;
          const createdAt = typeof value.createdAt === "string" ? Date.parse(value.createdAt) : NaN;
          if (!Number.isFinite(createdAt) || createdAt < sinceMs) continue;
          const parts = r.uri.replace("at://", "").split("/");
          const [parsedDid, parsedColl, parsedRkey] = parts;
          if (!parsedDid || !parsedColl || !parsedRkey) continue;
          yield {
            source: "catchup",
            did: parsedDid,
            collection: parsedColl,
            rkey: parsedRkey,
            uri: r.uri,
            cid: r.cid,
            timeMs: createdAt,
            record: value,
          };
        }
        cursor = res.data.cursor;
      } while (cursor);
    }
  }
}
